### PR TITLE
story: modeler extension

### PR DESCRIPTION
**Pull request by mbritense**
for https://ritense.tpondemand.com/entity/129202-febe-rework-process-management-screen

### CODE REVIEW

File: projects/valtimo/process-link/src/lib/components/process-link/process-link.component.ts [hunk 1]
```diff
@@ -43,6 +43,8 @@ export class ProcessLinkComponent {
 
     this.modalService.setModalData(params);
 
+    console.log('set modal data', params);
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/process-link/process-link.component.ts]: Remove log

File: projects/valtimo/process-link/src/lib/components/select-form-flow/select-form-flow.component.ts [hunk 2]
```diff
@@ -171,30 +180,34 @@ export class SelectFormFlowComponent implements OnInit, OnDestroy {
...
+        };
+
+        if (this.stateService.processLinkEditMode === ProcessLinkEditMode.EMIT_EVENTS) {
+          this.stateService.sendProcessLinkCreateEvent(createRequest);
+          return;
         }
-      );
+
+        this.processLinkService.saveProcessLink(createRequest).subscribe(
+          () => {
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/select-form-flow/select-form-flow.component.ts]: Write with, 'next' and 'error'

File: projects/valtimo/process-link/src/lib/components/select-form/select-form.component.ts [hunk 3]
```diff
@@ -177,35 +186,39 @@ export class SelectFormComponent implements OnInit, OnDestroy {
...
+        };
+
+        if (this.stateService.processLinkEditMode === ProcessLinkEditMode.EMIT_EVENTS) {
+          this.stateService.sendProcessLinkCreateEvent(createRequest);
+          return;
         }
-      );
+
+        this.processLinkService.saveProcessLink(createRequest).subscribe(
+          () => {
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/select-form/select-form.component.ts]: Write with 'next' and 'error'

File: projects/valtimo/process-link/src/lib/components/select-url/select-url.component.ts [hunk 4]
```diff
@@ -145,24 +131,29 @@ export class SelectUrlComponent implements OnInit, OnDestroy {
...
+          activityId: modalParams.element.id,
+        };
+
+        if (this.stateService.processLinkEditMode === ProcessLinkEditMode.EMIT_EVENTS) {
+          this.stateService.sendProcessLinkCreateEvent(processLinkRequest);
+          return;
+        }
+
+        this.processLinkService.saveProcessLink(processLinkRequest).subscribe(
+          () => {
```
Comment (teodor-ritense) [inline projects/valtimo/process-link/src/lib/components/select-url/select-url.component.ts]: Write with 'next' ad 'error'

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  *ngIf="{
+    processDefinitionVersions: processDefinitionVersions$ | async,
+    selectedProcessDefinitionXml: selectedProcessDefinitionXml$ | async,
+    loading: loading$ | async,
+    isReadOnlyProcess: isReadOnlyProcess$ | async,
+  } as obs"
+  data-carbon-theme="g10"
+  class="bpmn__container"
+>
+  <div *ngIf="obs.loading" class="bpmn__loading">
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html:29]: Can this go directly on the cds-loading?

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    <div class="bpmn__modeler-canvas" #modeler></div>
+
+    <div class="bpmn__modeler-panel" #modelerPanel></div>
+  </div>
+
+  <div class="bpmn__modeler" [style.display]="obs.isReadOnlyProcess ? 'flex' : 'none'">
+    <div class="bpmn__modeler-canvas" #viewer></div>
+
+    <div class="bpmn__modeler-panel" #viewerPanel></div>
+  </div>
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html:43]: Use @if

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+        loading: loading$ | async,
+        changesPending: changesPending$ | async,
+        isReadOnlyProcess: isReadOnlyProcess$ | async,
+        isSystemProcess: isSystemProcess$ | async,
+        compactMode: compactMode$ | async,
+        creatingNewProcess: creatingNewProcess$ | async,
+      } as actionsObs"
+    >
+      <div class="page-header-actions__left-container">
+        <div *ngIf="!actionsObs.creatingNewProcess" class="page-header-actions__version-dropdown">
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html]: Can this go directly on the cds-dropdown?

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        </div>
+      </div>
+
+      <div class="page-header-actions__buttons">
+        <cds-overflow-menu class="overflow-button">
+          <cds-overflow-menu-option
+            [disabled]="actionsObs.loading"
+            (selected)="export(actionsObs.isReadOnlyProcess)"
+            >{{ 'interface.export' | translate }}</cds-overflow-menu-option
+          >
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html]: Format this

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.scss [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    display: flex;
+    gap: 16px;
+  }
+}
+
+::ng-deep .process-link-properties-panel {
+  padding: 8px 12px 16px 12px;
+  width: 100%;
+  display: flex;
+  flex-direction: row;
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.scss]: Default direction is row

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.ts [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+
+  public ngOnDestroy(): void {
+    this._bpmnModeler?.destroy();
+    this._bpmnViewer?.destroy();
+    this._subscriptions.unsubscribe();
+  }
+
+  public deployChanges(isReadOnlyProcess: boolean): void {
+    combineLatest([
+      isReadOnlyProcess ? from(this._bpmnViewer.saveXML()) : from(this._bpmnModeler.saveXML()),
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.ts]:
```suggestion
      from(isReadonlyProcess ? this.bpmnViewer.saveXML() : this._bpmnModeler.saveXML()),
```

File: projects/valtimo/process-management/src/lib/components/process-management-list/process-management-list.component.ts [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    switchMap(() =>
+      this.processService.getProcessDefinitions().pipe(tap(() => this.loading$.next(false)))
+    )
+  );
+
+  public readonly FIELDS = [
+    {key: 'name', label: 'Name'},
+    {key: 'key', label: 'Key'},
+    {key: 'readOnly', label: 'Read-only'},
+  ] as ColumnConfig[];
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-list/process-management-list.component.ts]: Why not give it type ColumnConfig[] directly?

File: projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.html [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+  -->
+
+<cds-modal valtimoCdsModal [open]="modalOpen$ | async" size="sm">
+  <cds-modal-header showCloseButton="true" (closeSelect)="closeModal()">
+    <h3 cdsModalHeaderHeading>
+      <h3 class="modal-title">Upload process</h3>
+    </h3>
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.html]:
```suggestion
      <h3 cdsModalHeaderHeading class="modal-title">Upload process</h3>
```

File: projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.html [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+        type="button"
+        class="btn btn-primary btn-space"
+        (click)="uploadProcessBpmn()"
+        [disabled]="!bpmn"
+        data-dismiss="modal"
+      >
+        Upload
+      </button>
+    </div>
+  </cds-modal-footer>
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.html]: Put class btn-group directly on cds-modal-footer

File: projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.ts [hunk 14]
```diff
@@ -25,25 +26,32 @@ import {AlertService} from '@valtimo/components';
...
+  public closeModal(): void {
+    this.processManagementStateService.closeModal();
+  }
+
+  public onChange(files: FileList): void {
     this.bpmn = files.item(0);
   }
 
-  uploadProcessBpmn() {
+  public uploadProcessBpmn() {
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.ts]: Return type

File: projects/valtimo/process-management/src/lib/process-management.service.ts [hunk 15]
```diff
@@ -42,4 +44,29 @@ export class ProcessManagementService {
       formData
     );
   }
+
+  getProcessLinks(processDefinitionId: string) {
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/process-management.service.ts]:
```suggestion
  public getProcessLinks(processDefinitionId: string) {
```

File: projects/valtimo/process-management/src/lib/process-management.service.ts [hunk 16]
```diff
@@ -42,4 +44,29 @@ export class ProcessManagementService {
...
+  public processLinks$ = new BehaviorSubject<ProcessLink[]>([]);
+
+  constructor(
+    private readonly processLinkService: ProcessLinkService,
+    private readonly processDefinitionId: string
+  ) {
+    this.loadProcessLinks()
+  }
+
+  loadProcessLinks() {
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/process-management.service.ts]:
```suggestion
  public loadProcessLinks(): void {
```

File: src/app/app-routing.module.ts [hunk 17]
```diff
@@ -41,7 +41,7 @@ const routes: Routes = [
 export class AppRoutingModule {
   constructor(private router: Router) {
     this.router.errorHandler = (error: any) => {
-      this.router.navigate(['']);
+      console.log(error)
```
Comment (teodor-ritense) [inline src/app/app-routing.module.ts]: Maybe console.error?

Or do we wanna keep the navigation to home?

File: projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html (hunk 5 above)
Comment (mbritense) [reply projects/valtimo/process-management/src/lib/components/process-management-builder/process-management-builder.component.html:29]: No it's an overlay container

**Review by teodor-ritense — DISMISSED** [review]

File: projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.ts [hunk 18]
```diff
@@ -0,0 +1,86 @@
...
+          duration: 5000,
+        });
+        this.closeModal();
+        this.processManagementStateService.reloadDefinitions();
+      },
+      error: () => {
+        this.notificationService.showNotification({
+          type: 'error',
+          title: this.translateService.instant('processManagement.upload.failure'),
+          duration: 5000,
```
Comment (teodor-ritense) [inline projects/valtimo/process-management/src/lib/components/process-management-upload/process-management-upload.component.ts]: Use CARBON_CONSTANTS

**Review by teodor-ritense — APPROVED** [review]

**Pull request merged**